Disable Glish on Bare Metal Machines

bcmertz opened the pull request:
## Description

This disables glish for bare metal machines by checking if the current Linode is bare metal

## How to test

First you must authorize yourself for bare metal then create a bare metal linode and a non bare metal linode. Verify that the bare metal one has no glish and the non bare metal one does

Note: Not important but I went down a bunch of dead ends trying to use src/hooks/useExtendedLinode but because Lish is not a FC that didn't work. Then I thought about passing in the data from store.getState() in src/index.tsx because I misguidedly thought that perhaps Lish being a separate popup couldn't access the store and needed it passed in without realizing api calls work. TLDR - read documentation 🙃

bnussman commented on packages/manager/src/features/Lish/Lish.tsx:
```diff
@@ -241,7 +254,7 @@ class Lish extends React.Component<CombinedProps, State> {
               </SafeTabPanel>
             )}
-            {linode && token && (
+            {linode && token && isBareMetal && (
```
I think this should be `!isBareMetal`:

```ts
{linode && token && !isBareMetal && (
```

bcmertz replied:
good catch, must've mixed up my linodes while testing

bnussman commented on packages/manager/src/features/Lish/Lish.tsx:
```diff
@@ -209,6 +206,22 @@ class Lish extends React.Component<CombinedProps, State> {
+    if (!isBareMetal) {
+      this.tabs = [
+        /* NB: These must correspond to the routes inside the Switch */
+        {
+          title: 'Weblish',
+          routeName: `${this.props.match.url}/weblish`,
+        },
```
Instead of repeating this code for the weblish tab, we could potentially use a spread operator. Maybe something like:

```ts
let tabs = [
  {
    title: 'Weblish',
    routeName: `${this.props.match.url}/weblish`,
  },
];

if (!isBareMetal) {
  tabs = [
    ...tabs,
    {
      title: 'Glish',
      routeName: `${this.props.match.url}/glish`,
    },
  ];
}
```

bcmertz replied:
thanks! I think it might be nice to keep it typed like @DevDW suggested

bnussman reviewed (changes requested):
I'm still seeing Glish for my Bare Metal Linode 😧. I'm not sure exactly why that is but I'm thinking it may have something to do with tabs being a variable and not part of the components state. I had success with making the tabs variable a variable that is local to the render method.  I'm sure there are numerous way to go about this. 

As a side note for testing PR's like this, we could consider adding a Bare Metal Linode to the Mock Service Worker

DevDW reviewed (commented):
I'm observing the same thing as @bnussman and agree with most of his suggestions. Another option for the `tabs` variable after moving it into the `render()` method is:

```
    const tabs = [
      /* NB: These must correspond to the routes inside the Switch */
      {
        title: 'Weblish',
        routeName: `${this.props.match.url}/weblish`,
      },
      !isBareMetal
        ? {
            title: 'Glish',
            routeName: `${this.props.match.url}/glish`,
          }
        : null,
    ].filter(Boolean) as Tab[];
```

where the `Tab` used for the type is imported from `TabLinkList.tsx`.

Perhaps some styling tweaks are also in order when the Glish tab is hidden (bottom white border)?

<img width="1078" alt="Screen Shot 2021-09-13 at 12 47 28 PM" src="https://user-images.githubusercontent.com/32860776/133124524-1e81268a-4b38-4143-9445-962efdbd8954.png">

bcmertz commented:
> As a side note for testing PR's like this, we could consider adding a Bare Metal Linode to the Mock Service Worker

Great idea, I'll add that to this PR!

bcmertz commented:
> I'm observing the same thing as @bnussman and agree with most of his suggestions. Another option for the `tabs` variable after moving it into the `render()` method is:
> 
> ```
>     const tabs = [
>       /* NB: These must correspond to the routes inside the Switch */
>       {
>         title: 'Weblish',
>         routeName: `${this.props.match.url}/weblish`,
>       },
>       !isBareMetal
>         ? {
>             title: 'Glish',
>             routeName: `${this.props.match.url}/glish`,
>           }
>         : null,
>     ].filter(Boolean) as Tab[];
> ```
> 
> where the `Tab` used for the type is imported from `TabLinkList.tsx`.
> 
> Perhaps some styling tweaks are also in order when the Glish tab is hidden (bottom white border)?
> 
> <img alt="Screen Shot 2021-09-13 at 12 47 28 PM" width="1078" src="https://user-images.githubusercontent.com/32860776/133124524-1e81268a-4b38-4143-9445-962efdbd8954.png">

Done!

bnussman reviewed (approved):
Nice job addressing feedback! Looks good 💻

bcmertz commented:
> Nice job addressing feedback! Looks good 💻

thanks! I added a bare metal mock Linode, would you mind checking it? It appears the reason glish appears for the mocked bare metal is because api-v4/lib/linodes/linodes.js/getLinode 404s. Is this expected?

DevDW reviewed (approved):
Nice work! 🚢

bnussman commented:
That 404 would be expected. You'd have to add something like this below, where the id of the Linode `26` matches the id of the Bare Metal Linoide returned in the `*/linode/instances` mock.   

```ts
rest.get('*/instances/26', async (req, res, ctx) => {
  return res(
    ctx.json(
      linodeFactory.build({
        type: 'metal',
      })
    )
  );
}),
```

DevDW commented on packages/manager/src/features/Lish/Lish.tsx:
```diff
@@ -227,35 +233,40 @@ class Lish extends React.Component<CombinedProps, State> {
+    }
+
     return (
       <React.Fragment>
```
can add

`// eslint-disable-next-line react/jsx-no-useless-fragment`

above this line to make the linter happy:

<img width="973" alt="Screen Shot 2021-09-16 at 5 00 30 PM" src="https://user-images.githubusercontent.com/32860776/133685024-0eb4a821-b80e-44bd-8c0d-a9da25c919ee.png">
